### ISSUE TITLE
Include sniff codes in the docs when using --generator argument

### ISSUE THREAD

**Issue by przemekhernik** [PHPCSStandards/PHP_CodeSniffer#317]
## Is your feature request related to a problem?

Searching for proper sniffs/sniffs explanations was always a complicated thing when using `phpcs`, at least for me, especially as a newcomer. I've had several approaches over the years to this, without success, and now I've decided to give it a try one more. 

What problems I have before digging deeper in the phpcs codebase?

1. **Building own ruleset**: When I wanted to build my own ruleset I couldn't find an easy way to find the sniff codes that I should include. Some sniffs include docs, that can be generated with `phpcs --standard=Generic --generator=text` but the codes are not available there so when I found something useful, I had to use a little bit of reverse engineering to find out the code. 

2. **Getting sniffs explanations:** I know that we can get a sniff explanation (when docs exist) with `phpcs --standard=Generic --generator=text --sniffs=Generic.Metrics.CyclomaticComplexity` but it would be also nice to search for them easier, especially for newcomers. 

When I was searching for solutions I found a few discussions with similar concerns. That's why I decided to ask for this.

- https://github.com/squizlabs/PHP_CodeSniffer/issues/1752#issuecomment-750927206

## Describe the solution you'd like

It would be much easier for me as a beginner if the sniff code were available in the docs generated with `--generate` attribute. So I could just generate the docs for specific standards, select the sniffs that I'm interested in, and take the sniff codes easily. Of course, it's not impossible to find out the code, but such a simple improvement would be nice.

I've already decided to [give it a try for my internal needs](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/312/files) and It was really useful! The changes/propositions:
- [Display sniff code](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/312/commits/e3ddb98da1e0f95bb08bfefbaa32f2f7693cf307) in the documentation created with `--generate` tag.
- Allow to [set the sniff code](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/312/commits/5f1b02f1d5a373ddacdfcbe4ef7ea231e9a920dc) in the `code` attribute of the `documentation` XML tag (optional). 
- When the sniff code is not set in the mentioned tag, it is [automatically detected](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/312/commits/0bc8a3f1b6e8bf47745211b251b87c08a1d5d88d) by phpcs engine [and set](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/312/commits/1c957f748734a23b466a9ae1a2185626b9826e0a). 

Thanks to this I'm able to:
- Review the generated docs and find the sniff codes to include/exclude them in the ruleset much faster.
- Find more information about errors much easier by searching info by sniff code.

I'm aware that I might miss something crucial that would similarly solve those problems, but if not, maybe it will be useful for others. What do you think?

#### HTML
<img width="762" alt="HTML" src="https://github.com/PHPCSStandards/PHP_CodeSniffer/assets/13350922/276f5970-52f9-4401-910c-9351f0008d2c">

#### Text
<img width="890" alt="Text" src="https://github.com/PHPCSStandards/PHP_CodeSniffer/assets/13350922/c6fb326a-f563-49c8-8968-c14b7d207340">

#### Markdown
<img width="1005" alt="markdown" src="https://github.com/PHPCSStandards/PHP_CodeSniffer/assets/13350922/f7789cde-3a8b-4031-a355-3e8fee01fd27">

## Additional context (optional)

- [x] I intend to create a pull request to implement this feature.

**Comment by jrfnl** [comment]
Thanks @przemekhernik for opening this ticket.

I understand what you are saying and can understand that this would seem like a quick fix, but I think it would be prudent to look at the docs format in a more holistic manner.

What I mean by that, is that I think the docs feature should be looked at not just for how it is available and used now, but what would be a way to make this feature more useful to end-users in general.

Some ideas I have floating around about this:
* Document the error code(s) each `<standard>` block covers.
    The ticket description above talks about the three part sniff codes - `Stdn.Category.SniffName`, but every error/warning also has a fourth part - `Stnd.Category.SniffName.ErrorCode` and generally speaking each `<standard>` block + code samples covers one or more error codes.
    As error codes can be included/excluded individually from a custom ruleset, I believe it could be useful make the error codes which are used by a sniff available via the XML Docs.
* Document public properties which can be set on a sniff and what they do.
    These are currently [documented in the wiki](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties) and that list is maintained manually and does not have any direct link with the XML docs.
* Allow for reference URLs to be added to XML docs, either for the whole sniff and/or for individual "standards" within the docs.
    Think: links to the FIG PSR2/PSR12 specifications for the PSR2/PSR12 sniffs.
    But also, links to the PHP manual or to a PHP RFC for the PHPCompatibility sniffs, or to a company coding standards manual for company-specific sniffs, or to a CMS/Framework manual for CMS/Framework specific sniffs.

The problem with any and all of these changes, would be that any change to a sniff _may_ cause the docs to become out of date, so this would add a maintenance and review burden and this needs some thinking about if and how this could be (partially) automated.

Some automation ideas:
* Automatically run the sniff over the valid/invalid code samples in the associated XML docs to verify that these still line up.
* Scan the sniff code for all `addError()`/`addWarning()` calls, gather the error codes/error code patterns and verify against error codes mentioned in the docs.
* Use reflection on the sniff class to verify that all public properties in the sniff are mentioned in the docs and that no public properties are mentioned in the docs which no longer exist on a sniff.
* Check URLs mentioned in the docs do not 404.
* Possibly automate generating/updating the wiki page based on the XML docs.
* For all these automation ideas, it should be considered whether the automation could also be useful for external standards with XML docs for their sniffs and if so, it should be set up to allow for them to use it too.
    As an example, [PHPCSDevTools](https://github.com/PHPCSStandards/PHPCSDevTools) currently already contains two tools related to the XML docs - an XSD file to validate XML docs against and a "check complete" tool which can check that all sniffs are accompanied by tests and documentation.

Aside from the information to be contained in the docs, I also think any discussion about changes to the docs format needs to be accompanied by a discussion on improving the _discoverability_ of sniffs. 

Currently there are three PHPCS native generators: Text, HTML, Markdown.
I don't know if there are other custom generators in use in external standards. It would be nice to get some insight into that.

While you clearly have discovered the Docs generator feature, I have a feeling that the majority of users don't know it exists and don't use it, so what would be a way to make this wealth of information more easily discoverable by end-users ?

One idea I've been thinking about is to add a generator which would parse the docs to a format which could be used by Jekyll based GH Pages and to then publish a website with the documentation of all sniffs, including a search feature.
I can even imagine, including the docs of select external standards in this search feature as well, so one could search on "import use" and see which sniffs are available related to import `use` statements in various PHPCS native and external standards to make an informed choice.

Aside from the initial time investments setting this up, this again would need to be accompanied by automation to automatically update the website on a new release of PHPCS itself and on releases of any external standards for which docs are included.

However, all of this is only useful if there _are_ docs available for all sniffs, which is why I opened issue #148 to at least try to get XML docs in place for all PHPCS native sniffs.
Similar tickets are open in some external standards I'm involved with, like [PHPCompatibility](https://github.com/PHPCompatibility/PHPCompatibility/issues/1285). (not so in [PHPCSExtra](https://github.com/PHPCSStandards/PHPCSExtra) as that has had a docs _requirement_ from the start, so all sniffs already have docs :tada:)

Priority wise, changing the docs format is not high on my list at this time as there are more urgent things to address in PHPCS.

First priority regarding the docs now, is [getting the docs for each sniff in place](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/148) and there is still enough work to do there.
After that we can look at enhancing the available information.

Having said that, I do think it would be good to get a discussion going about this topic, so that when the time is right, all ideas and opinions have been gathered and we can take informed decisions to move this forward.

As part of this ticket, I think it would also be good to gather a list of tickets related to this topic from the Squizlabs repo as I know there have been a few discussions related to this in the past.

Either way, that's my two pennies for now. None of this is set in stone, these are just ideas and other ideas, either for features to add to the XML docs or for ways to make the docs more useful/discoverable are very welcome!

**Comment by jrfnl** [comment]
> if you consider my initial idea about putting the sniff codes into the docs as useful too, maybe we can think about implementing this now?

I rather not. Why would your "random" change get priority over changes which have been requested by others ? I think any changes need a good think about the whole thing, including what changes are needed to the XSD and how to keep things readable without too much noise.

**Comment by przemekhernik** [comment]
> Why would your "random" change get priority over changes to the docs which have been requested by others ?

Oh, I see. I just wanted to help with this implementation so I thought that if there's a possibility to improve it a little bit, pretty easily, it would be fine to go for it to make the next step to make the tool better. But If you consider this as a random change, not a needed improvement, I understand that it's not worth discussing it at least now 🫡

Let's skip this and take care of things you suggested 😀

**Comment by jrfnl** [comment]
@przemekhernik Don't get me wrong, I do think it is a valuable suggestion and a valid request.
I just don't think that it should be prioritized over improving the docs in a more holistic way.

**Comment by przemekhernik** [comment]
Thanks for saying this! I'll try to think about the things you've mentioned to make the docs better 🤝

Meantime, I've created a simple bash script that iterates through the default standards and shows sniff codes with explanations in one place. It will help me or others solve the problems I mentioned earlier and focus on internal docs improvements. That's not perfect, but at least it's easier to find specific sniffs 🙏

It is available here if someone want to check it: https://tentyp.dev/library/php/phpcs-cheatset/ 👋

**Comment by jrfnl** [comment]
@przemekhernik Nice! Can I recruit you to help with getting a website for PHPCS set up ? 😉

**Comment by jrfnl** [comment]
@fredden 

> I've had a look into this and can see how this can be achieved within the current test-suite, without too much difficulty. The main issue is that we currently have no way of knowing which error codes are covered by the documentation. When we have this information available in the XSD, we can add this functionality if we still want it.

Nice! I would prefer a check separate from the test suite though, as while it is about QA, it is not a _test_. I imagine it could be a tool to be added to [PHPCSDevTools](https://github.com/PHPCSStandards/PHPCSDevTools).

> While looking into this, I noticed that it would be trivial to add a warning if there is no XML documentation file for a given sniff

That is also something which should not be part of the test suite and there is already a separate tool available for this, which will be added and turned on once the docs are complete: https://github.com/PHPCSStandards/PHPCSDevTools#checking-whether-all-sniffs-in-a-phpcs-standard-are-feature-complete

> and if there are fixable errors/warnings, but no .fixed file.

See issue #299 + #300 and in particular https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/300#issuecomment-1905144488
Short version: I have a commit ready for this already, but this change will need to go into 4.0.

**Comment by przemekhernik** [comment]
> Nice! Can I recruit you to help with getting a website for PHPCS set up ? 😉

We need to discuss it with my wife and toddler 🤣 But sure, I can try to help in this area. But at first, I'll try to think about what can we do to [have all the docs in place](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/148).

Currently, I'm working on a video material related to using `phpcs` in the development process. There were tons of such materials already, but I wanted to focus on the things that were less covered in them like the docs discoverability. I believe that maybe with this we'll be able to find more people to help with the docs. We just need about 56 sniffs to be documented to have them all 😅 I hope this material will contribute to the project by finding more people who are willing to help 🙏

I'll be recording probably next week, so I'll be pleased to hear your thoughts about the things that were hard - at least for me - the configuration. The general thoughts are [here](https://tentyp.dev/blog/phpcs/#configuration). Of course just only if you have a time!

**Comment by jrfnl** [comment]
> The general thoughts are [here](https://tentyp.dev/blog/phpcs/#configuration). Of course just only if you have a time!

@przemekhernik I've had a quick read-through and would make the following suggestions:
* Example command contains a typo: `--colored` should be `--colors`.
* I'd recommend not using `-n` in the examples as that's hiding all warnings and new users will follow advise like that blindly and not realize they are silencing half the information.
* I'd recommend calling the config file `phpcs.xml.dist` or `.phpcs.xml.dist`, not `phpcs.xml` (That way, devs can overload the project ruleset with their own version in a `phpcs.xml` file, which includes the dist file).
* I'd recommend using selective ignores in the `// phpcs:ignore` example as a plain `// phpcs:ignore` will ignore the complete line, while a selective ignore `// phpcs:ignore Stnd.Cat.Sniff.Errorcode` will ignore only that which should be ignored and will still show other errors when the line is updated.
    Or at the very least, mention that ignores/disables etc can be made selective.
* Also please fix the typo in my GitHub handle (in multiple places).

**Comment by stronk7** [comment]
Just coming from #148 (let's add all the missing docs), I think that there are various aspects that may be documented (basically all already named here):
- error codes.
- configurable properties.
- custom config settings.

And, maybe, we should consider not only having docs for the individual Sniffs, but also for their categories (namespace tree) and for the standard itself.

About how much of that can be automated, ideally being able to do it within code would be phenomenal (within phpdoc block or so). Keeping the docs manually maintained (into structured format, like current Sniff docs, or manually, like current property customisations wiki) is a real pain.

To dream is cheap™ , great work everybody!

Ciao :-)

**Comment by braindawg** [comment]
Just popping in to say that #312 would have saved me a good deal of time today, for what that's worth. I recognize that it's likely a small subset of users that are aware of the doc generators at all, but adding the sniff names seems like an easy win without taking away from any of the other doc-related initiatives. Thanks, everyone, for all your work on this project!

**Comment by yCodeTech** [comment]
It would be sooo useful to have a full list of all the sniffs and params to look through on GitHub. I know there is the [Customisable Sniff Properties](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties) page in the wiki, but this is only a subset of the sniffs. A lot of us exclude or include certain sniffs in our projects, but there's not many ways we can obtain the sniff without it being errored by phpcs beforehand.

I'm trying to find all the sniffs relating to commenting and docblocks. The wiki only mentions a limited amount of customisable sniffs. I am using a vscode extension to automatically fix any fixable issues on save of the file; and I can only find out what sniffs are in use when they are flagged up by the vscode inline error reporting.

Please please, really consider adding a full list of all sniffs to the wiki, so we can easily find, copy and paste them into our config to include/exclude them. This would be really helpful.

**Comment by yCodeTech** [comment]
@jrfnl 

> Something along those lines is planned for the website when it eventually becomes available.

Ah ok.

> In the mean time... are you aware you can use the `-e` option to discover what sniffs are available ? Even if you don't use a particular standard, as long as it is installed, you can use `-e` with the `--standard=..` flag to see what sniffs are available via the standard.

Yeah I've managed to do something with that. I've actually created a script to get all the sniffs that are available from all the installed standards. The script uses the `-i` option first to get a list of all the installed standards, and then uses the `-e` option with with the `--standard=[list of installed standards]` as you've mentioned to get the sniffs. I have then parsed that output and generated an associative array with all the sniffs grouped into their respective standards and also separated the deprecated sniffs in each standard too. This is a stepping stone in what could be a good generator.

The script is a bit rudimentary at the moment, but I've put together a [repo for it](https://github.com/yCodeTech/list-phpcs-sniffs) with the [script in question](https://github.com/yCodeTech/list-phpcs-sniffs/blob/master/ListPhpCsSniffs.php).

I think this script could potentially work with this repo somehow. I'd be willing to collab on this if necessary.

This is my list of sniffs:
<details><summary>List of sniffs</summary>


```
array(7) {
  ["MySource"]=>
  array(1) {
    ["deprecated"]=>
    array(16) {
      [0]=>
      string(37) "MySource.Channels.DisallowSelfActions"
      [1]=>
      string(34) "MySource.Channels.IncludeOwnSystem"
      [2]=>
      string(31) "MySource.Channels.IncludeSystem"
      [3]=>
      string(30) "MySource.Channels.UnusedSystem"
      [4]=>
      string(35) "MySource.Commenting.FunctionComment"
      [5]=>
      string(34) "MySource.CSS.BrowserSpecificStyles"
      [6]=>
      string(24) "MySource.Debug.DebugCode"
      [7]=>
      string(29) "MySource.Debug.FirebugConsole"
      [8]=>
      string(27) "MySource.Objects.AssignThis"
      [9]=>
      string(41) "MySource.Objects.CreateWidgetTypeCallback"
      [10]=>
      string(34) "MySource.Objects.DisallowNewWidget"
      [11]=>
      string(31) "MySource.PHP.AjaxNullComparison"
      [12]=>
      string(30) "MySource.PHP.EvalObjectFactory"
      [13]=>
      string(27) "MySource.PHP.GetRequestData"
      [14]=>
      string(32) "MySource.PHP.ReturnFunctionValue"
      [15]=>
      string(28) "MySource.Strings.JoinStrings"
    }
  }
  ["PEAR"]=>
  array(18) {
    [0]=>
    string(41) "PEAR.ControlStructures.MultiLineCondition"
    [1]=>
    string(24) "PEAR.Files.IncludingFile"
    [2]=>
    string(35) "PEAR.Formatting.MultiLineAssignment"
    [3]=>
    string(36) "PEAR.Functions.FunctionCallSignature"
    [4]=>
    string(32) "PEAR.Functions.ValidDefaultValue"
    [5]=>
    string(29) "PEAR.Classes.ClassDeclaration"
    [6]=>
    string(28) "PEAR.Commenting.ClassComment"
    [7]=>
    string(27) "PEAR.Commenting.FileComment"
    [8]=>
    string(31) "PEAR.Commenting.FunctionComment"
    [9]=>
    string(29) "PEAR.Commenting.InlineComment"
    [10]=>
    string(39) "PEAR.ControlStructures.ControlSignature"
    [11]=>
    string(34) "PEAR.Functions.FunctionDeclaration"
    [12]=>
    string(37) "PEAR.NamingConventions.ValidClassName"
    [13]=>
    string(40) "PEAR.NamingConventions.ValidFunctionName"
    [14]=>
    string(40) "PEAR.NamingConventions.ValidVariableName"
    [15]=>
    string(36) "PEAR.WhiteSpace.ObjectOperatorIndent"
    [16]=>
    string(33) "PEAR.WhiteSpace.ScopeClosingBrace"
    [17]=>
    string(27) "PEAR.WhiteSpace.ScopeIndent"
  }
  ["PSR2"]=>
  array(12) {
    [0]=>
    string(32) "PSR2.Classes.PropertyDeclaration"
    [1]=>
    string(25) "PSR2.Files.EndFileNewline"
    [2]=>
    string(30) "PSR2.Methods.MethodDeclaration"
    [3]=>
    string(29) "PSR2.Classes.ClassDeclaration"
    [4]=>
    string(46) "PSR2.ControlStructures.ControlStructureSpacing"
    [5]=>
    string(40) "PSR2.ControlStructures.ElseIfDeclaration"
    [6]=>
    string(40) "PSR2.ControlStructures.SwitchDeclaration"
    [7]=>
    string(21) "PSR2.Files.ClosingTag"
    [8]=>
    string(34) "PSR2.Methods.FunctionCallSignature"
    [9]=>
    string(33) "PSR2.Methods.FunctionClosingBrace"
    [10]=>
    string(36) "PSR2.Namespaces.NamespaceDeclaration"
    [11]=>
    string(30) "PSR2.Namespaces.UseDeclaration"
  }
  ["Squiz"]=>
  array(74) {
    ["deprecated"]=>
    array(25) {
      [0]=>
      string(31) "Squiz.Classes.DuplicateProperty"
      [1]=>
      string(42) "Squiz.CSS.ClassDefinitionClosingBraceSpace"
      [2]=>
      string(36) "Squiz.CSS.ClassDefinitionNameSpacing"
      [3]=>
      string(42) "Squiz.CSS.ClassDefinitionOpeningBraceSpace"
      [4]=>
      string(22) "Squiz.CSS.ColonSpacing"
      [5]=>
      string(26) "Squiz.CSS.ColourDefinition"
      [6]=>
      string(42) "Squiz.CSS.DisallowMultipleStyleDefinitions"
      [7]=>
      string(34) "Squiz.CSS.DuplicateClassDefinition"
      [8]=>
      string(34) "Squiz.CSS.DuplicateStyleDefinition"
      [9]=>
      string(30) "Squiz.CSS.EmptyClassDefinition"
      [10]=>
      string(30) "Squiz.CSS.EmptyStyleDefinition"
      [11]=>
      string(25) "Squiz.CSS.ForbiddenStyles"
      [12]=>
      string(21) "Squiz.CSS.Indentation"
      [13]=>
      string(34) "Squiz.CSS.LowercaseStyleDefinition"
      [14]=>
      string(22) "Squiz.CSS.MissingColon"
      [15]=>
      string(22) "Squiz.CSS.NamedColours"
      [16]=>
      string(17) "Squiz.CSS.Opacity"
      [17]=>
      string(26) "Squiz.CSS.SemicolonSpacing"
      [18]=>
      string(23) "Squiz.CSS.ShorthandSize"
      [19]=>
      string(26) "Squiz.Debug.JavaScriptLint"
      [20]=>
      string(18) "Squiz.Debug.JSLint"
      [21]=>
      string(39) "Squiz.Objects.DisallowObjectStringIndex"
      [22]=>
      string(31) "Squiz.Objects.ObjectMemberComma"
      [23]=>
      string(41) "Squiz.WhiteSpace.LanguageConstructSpacing"
      [24]=>
      string(37) "Squiz.WhiteSpace.PropertyLabelSpacing"
    }
    [0]=>
    string(32) "Squiz.Arrays.ArrayBracketSpacing"
    [1]=>
    string(29) "Squiz.Arrays.ArrayDeclaration"
    [2]=>
    string(30) "Squiz.Classes.ClassDeclaration"
    [3]=>
    string(27) "Squiz.Classes.ClassFileName"
    [4]=>
    string(36) "Squiz.Classes.LowercaseClassKeywords"
    [5]=>
    string(33) "Squiz.Classes.SelfMemberReference"
    [6]=>
    string(28) "Squiz.Classes.ValidClassName"
    [7]=>
    string(29) "Squiz.Commenting.BlockComment"
    [8]=>
    string(29) "Squiz.Commenting.ClassComment"
    [9]=>
    string(42) "Squiz.Commenting.ClosingDeclarationComment"
    [10]=>
    string(36) "Squiz.Commenting.DocCommentAlignment"
    [11]=>
    string(34) "Squiz.Commenting.EmptyCatchComment"
    [12]=>
    string(28) "Squiz.Commenting.FileComment"
    [13]=>
    string(40) "Squiz.Commenting.FunctionCommentThrowTag"
    [14]=>
    string(30) "Squiz.Commenting.InlineComment"
    [15]=>
    string(44) "Squiz.Commenting.LongConditionClosingComment"
    [16]=>
    string(37) "Squiz.Commenting.PostStatementComment"
    [17]=>
    string(32) "Squiz.Commenting.VariableComment"
    [18]=>
    string(40) "Squiz.ControlStructures.ControlSignature"
    [19]=>
    string(41) "Squiz.ControlStructures.ElseIfDeclaration"
    [20]=>
    string(46) "Squiz.ControlStructures.ForEachLoopDeclaration"
    [21]=>
    string(42) "Squiz.ControlStructures.ForLoopDeclaration"
    [22]=>
    string(43) "Squiz.ControlStructures.InlineIfDeclaration"
    [23]=>
    string(44) "Squiz.ControlStructures.LowercaseDeclaration"
    [24]=>
    string(41) "Squiz.ControlStructures.SwitchDeclaration"
    [25]=>
    string(25) "Squiz.Files.FileExtension"
    [26]=>
    string(32) "Squiz.Formatting.OperatorBracket"
    [27]=>
    string(35) "Squiz.Functions.FunctionDeclaration"
    [28]=>
    string(50) "Squiz.Functions.FunctionDeclarationArgumentSpacing"
    [29]=>
    string(41) "Squiz.Functions.FunctionDuplicateArgument"
    [30]=>
    string(30) "Squiz.Functions.GlobalFunction"
    [31]=>
    string(41) "Squiz.Functions.LowercaseFunctionKeywords"
    [32]=>
    string(44) "Squiz.Functions.MultiLineFunctionDeclaration"
    [33]=>
    string(41) "Squiz.NamingConventions.ValidFunctionName"
    [34]=>
    string(41) "Squiz.NamingConventions.ValidVariableName"
    [35]=>
    string(33) "Squiz.Objects.ObjectInstantiation"
    [36]=>
    string(39) "Squiz.Operators.ComparisonOperatorUsage"
    [37]=>
    string(39) "Squiz.Operators.IncrementDecrementUsage"
    [38]=>
    string(37) "Squiz.Operators.ValidLogicalOperators"
    [39]=>
    string(26) "Squiz.PHP.CommentedOutCode"
    [40]=>
    string(34) "Squiz.PHP.DisallowBooleanStatement"
    [41]=>
    string(38) "Squiz.PHP.DisallowComparisonAssignment"
    [42]=>
    string(26) "Squiz.PHP.DisallowInlineIf"
    [43]=>
    string(37) "Squiz.PHP.DisallowMultipleAssignments"
    [44]=>
    string(38) "Squiz.PHP.DisallowSizeFunctionsInLoops"
    [45]=>
    string(30) "Squiz.PHP.DiscouragedFunctions"
    [46]=>
    string(21) "Squiz.PHP.EmbeddedPhp"
    [47]=>
    string(14) "Squiz.PHP.Eval"
    [48]=>
    string(23) "Squiz.PHP.GlobalKeyword"
    [49]=>
    string(17) "Squiz.PHP.Heredoc"
    [50]=>
    string(24) "Squiz.PHP.InnerFunctions"
    [51]=>
    string(31) "Squiz.PHP.LowercasePHPFunctions"
    [52]=>
    string(27) "Squiz.PHP.NonExecutableCode"
    [53]=>
    string(26) "Squiz.Scope.MemberVarScope"
    [54]=>
    string(23) "Squiz.Scope.MethodScope"
    [55]=>
    string(27) "Squiz.Scope.StaticThisUsage"
    [56]=>
    string(34) "Squiz.Strings.ConcatenationSpacing"
    [57]=>
    string(30) "Squiz.Strings.DoubleQuoteUsage"
    [58]=>
    string(27) "Squiz.Strings.EchoedStrings"
    [59]=>
    string(28) "Squiz.WhiteSpace.CastSpacing"
    [60]=>
    string(40) "Squiz.WhiteSpace.ControlStructureSpacing"
    [61]=>
    string(42) "Squiz.WhiteSpace.FunctionClosingBraceSpace"
    [62]=>
    string(42) "Squiz.WhiteSpace.FunctionOpeningBraceSpace"
    [63]=>
    string(32) "Squiz.WhiteSpace.FunctionSpacing"
    [64]=>
    string(39) "Squiz.WhiteSpace.LogicalOperatorSpacing"
    [65]=>
    string(33) "Squiz.WhiteSpace.MemberVarSpacing"
    [66]=>
    string(38) "Squiz.WhiteSpace.ObjectOperatorSpacing"
    [67]=>
    string(32) "Squiz.WhiteSpace.OperatorSpacing"
    [68]=>
    string(34) "Squiz.WhiteSpace.ScopeClosingBrace"
    [69]=>
    string(36) "Squiz.WhiteSpace.ScopeKeywordSpacing"
    [70]=>
    string(33) "Squiz.WhiteSpace.SemicolonSpacing"
    [71]=>
    string(38) "Squiz.WhiteSpace.SuperfluousWhitespace"
    [72]=>
    string(32) "Squiz.Commenting.FunctionComment"
  }
  ["Zend"]=>
  array(3) {
    ["deprecated"]=>
    array(1) {
      [0]=>
      string(23) "Zend.Debug.CodeAnalyzer"
    }
    [0]=>
    string(21) "Zend.Files.ClosingTag"
    [1]=>
    string(40) "Zend.NamingConventions.ValidVariableName"
  }
  ["PSR1"]=>
  array(20) {
    [0]=>
    string(29) "PSR1.Classes.ClassDeclaration"
    [1]=>
    string(22) "PSR1.Files.SideEffects"
    [2]=>
    string(32) "PSR1.Methods.CamelCapsMethodName"
    [3]=>
    string(34) "PSR12.Classes.AnonClassDeclaration"
    [4]=>
    string(32) "PSR12.Classes.ClassInstantiation"
    [5]=>
    string(26) "PSR12.Classes.ClosingBrace"
    [6]=>
    string(31) "PSR12.Classes.OpeningBraceSpace"
    [7]=>
    string(48) "PSR12.ControlStructures.BooleanOperatorPlacement"
    [8]=>
    string(47) "PSR12.ControlStructures.ControlStructureSpacing"
    [9]=>
    string(28) "PSR12.Files.DeclareStatement"
    [10]=>
    string(22) "PSR12.Files.FileHeader"
    [11]=>
    string(27) "PSR12.Files.ImportStatement"
    [12]=>
    string(19) "PSR12.Files.OpenTag"
    [13]=>
    string(39) "PSR12.Functions.NullableTypeDeclaration"
    [14]=>
    string(37) "PSR12.Functions.ReturnTypeDeclaration"
    [15]=>
    string(36) "PSR12.Keywords.ShortFormTypeKeywords"
    [16]=>
    string(39) "PSR12.Namespaces.CompoundNamespaceDepth"
    [17]=>
    string(31) "PSR12.Operators.OperatorSpacing"
    [18]=>
    string(35) "PSR12.Properties.ConstantVisibility"
    [19]=>
    string(27) "PSR12.Traits.UseDeclaration"
  }
  ["JPSR12"]=>
  array(6) {
    [0]=>
    string(30) "JPSR12.Arrays.ArrayDeclaration"
    [1]=>
    string(37) "JPSR12.ControlSignatures.MultiKeyword"
    [2]=>
    string(22) "JPSR12.Files.BlankLine"
    [3]=>
    string(23) "JPSR12.Files.FileHeader"
    [4]=>
    string(32) "JPSR12.Operators.OperatorSpacing"
    [5]=>
    string(31) "JPSR12.Types.DeclareStrictTypes"
  }
}

```


</details>
